#576: Add support for AMMClawback

agutierrez0 opened the pull request:
closes #576 

- add `AmmClawback` to `SignatureUtils`, `Transaction.java`, `TransactionType.java`, and `TransactionTypeTests.java`
  - add test in `SignatureUtils` for the new `AmmClawback` class
- create `AmmClawbackFlags` with only `tfCalwTwoAssets` flag or unset flags
- create `AmmClawback` under `transactions`
- generate new `definitions.json`
- create new `AmmClawbackTest` transaction test
- updated rippled container version to `2.3.0` from `2.2.0-rc3`
- enable new `AMMClawback` amendment in `rippled.cfg`
- added integration tests under `AmmIt`
  - one test where we clawback with amount
  - one test where we clawback two assets using tfClawbackTwoAssets
  - one test where we clawback without amount and only clawback one asset

agutierrez0 commented on xrpl4j-core/src/main/java/org/xrpl/xrpl4j/model/flags/AmmClawbackFlags.java:
```diff
@@ -0,0 +1,51 @@
+   *
+   * @return {@code true} if {@code tfLPToken} is set, otherwise {@code false}.
+   */
+  public boolean tfClawTwoAssets() {
+    return this.isSet(CLAW_TWO_ASSETS);
+  }
```
This is unused, i added it to follow the practice I saw with other TransactionFlag classes I saw (e.g. `AmmWithdrawFlags`). Let me know if I should remove.

nkramer44 replied:
Let's keep it so developers can call `tfClawTwoAssets` rather than having to know to call `isSet(CLAW_TWO_ASSETS)`

agutierrez0 commented on xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/environment/RippledContainer.java:
```diff
@@ -84,7 +82,7 @@ public class RippledContainer {
    */
   public RippledContainer() {
-    try (GenericContainer<?> container = new GenericContainer<>("rippleci/rippled:2.2.0")) {
+    try (GenericContainer<?> container = new GenericContainer<>("rippleci/rippled:2.3.0")) {
```
Let me know if there is another version I could use here instead of `2.3.0`

nkramer44 replied:
I think version 2.3.1 is the most recent version, so we should use that

agutierrez0 commented on xrpl4j-core/src/main/java/org/xrpl/xrpl4j/model/transactions/AmmClawback.java:
```diff
@@ -0,0 +1,74 @@
+   *
+   * @return {@link AmmClawbackFlags#UNSET} if field was not set, otherwise returns with the set flag.
+   */
+  @JsonProperty("Flags")
+  @Value.Default
+  default AmmClawbackFlags flags() {
+    return AmmClawbackFlags.UNSET;
+  }
```
I noticed there was only one flag `AMMClawback` can be set to, this `tfClawTwoAssets` so I set the default to be `UNSET`, but if I should instead just make this an `Optional<>` field, let me know.

nkramer44 replied:
To align with other `Transaction` sub-types, this should default to `AmmClawbackFlags.empty()`

agutierrez0 commented on xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AmmIT.java:
For the three integration tests that I wrote, i would say around 90% of the code is the exact same:
1. create trust lines
2. send payments
3. create amm
4. deposit in amm
5. clawback assets in AMM

the only difference really is in the flags and/or the inclusion of `amount`. 

should I break down that that 90% into something more reusable? 

e.g. 

- create the following function`createTrustline`, `createPayment`, `depositToAmm` etc..
- adjust the `createAmm` function so that it can be reused

i had started going down this path, but wanted to get an opinion first.

sappenin commented on xrpl4j-core/src/main/java/org/xrpl/xrpl4j/model/transactions/AmmClawback.java:
```diff
@@ -0,0 +1,74 @@
+import org.xrpl.xrpl4j.model.flags.AmmClawbackFlags;
+import org.xrpl.xrpl4j.model.flags.TransactionFlags;
+import org.xrpl.xrpl4j.model.ledger.Issue;
+
```
```suggestion
import org.xrpl.xrpl4j.model.transactions.Address;
```

agutierrez0 replied:
i have added this extra import, but on my IDE it shows as unused.  

when i define the type for `holder`, i do see the `Address` references this path, but doesn't add the import. is there any reason why this happens?

agutierrez0 replied:
i have removed the extra import 👍🏽

sappenin commented on xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AmmIT.java:
```diff
@@ -288,6 +293,562 @@ void depositAndWithdraw() throws JsonRpcClientErrorException, JsonProcessingExce
+    // create Trustline from issuer to trader for currency TST and define two currencies
+    String testCurrency = "TST";
+    TrustSet trustSet = TrustSet.builder()
+        .account(traderKeyPair.publicKey().deriveAddress())
```
Need to fix formatting - use 2 spaces (no tabs)

agutierrez0 replied:
after finally being able to build, i have been able to address all of the checkstyle errors

sappenin commented on xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AmmIT.java:
```diff
@@ -288,6 +293,562 @@ void depositAndWithdraw() throws JsonRpcClientErrorException, JsonProcessingExce
+    assertThat(trustlineSubmitResult2.engineResult()).isEqualTo(TransactionResultCodes.TES_SUCCESS);
+
+    // send payment of ZFT to trader wallet from issuer
+    Payment paymentZFT = Payment.builder()
```
This naming convention breaks our Checkstyle plugin. Need to replace all instances of `...ZFT` with `Zft`. For example:

```suggestion
    Payment paymentZft = Payment.builder()
```

agutierrez0 replied:
fixed

sappenin commented on xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AmmIT.java:
```diff
@@ -288,6 +293,562 @@ void depositAndWithdraw() throws JsonRpcClientErrorException, JsonProcessingExce
+    assertThat(paymentZFTSubmitResult.engineResult()).isEqualTo(TransactionResultCodes.TES_SUCCESS);
+
+    // send payment of TST to trader wallet from issuer
+    Payment paymentTST = Payment.builder()
```
This naming convention breaks our Checkstyle plugin. Need to replace all instances of `...TFT` with `Tft`. For example:

```suggestion
    Payment paymentTft = Payment.builder()
```

agutierrez0 replied:
i've updated this along with all the other checkstyle issues

sappenin commented on xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AmmIT.java:
```diff
@@ -288,6 +293,562 @@ void depositAndWithdraw() throws JsonRpcClientErrorException, JsonProcessingExce
+    SubmitResult<TrustSet> trustlineSubmitResult2 = xrplClient.submit(signedTrustline2);
+    assertThat(trustlineSubmitResult2.engineResult()).isEqualTo(TransactionResultCodes.TES_SUCCESS);
+
+    // send payment of ZFT to trader wallet from issuer
```
Consider using real currency names, especially ones that are a familiar, like `usd` and `xrp`. It will be one less thing for developers to have keep track of while they seek to understand these tests.

agutierrez0 replied:
upd: build no longer failing

sappenin commented on xrpl4j-core/src/test/java/org/xrpl/xrpl4j/model/transactions/AmmClawbackTest.java:
```diff
@@ -0,0 +1,136 @@
+        .build();
+
+    String json = "{\n" +
+        "    \"Account\" : \"rJVUeRqDFNs2xqA7ncVE6ZoAhPUoaJJSQm\",\n" +
```
Remove un-necessary indents in JSON strings. For example: 
```suggestion
        "  \"Account\" : \"rJVUeRqDFNs2xqA7ncVE6ZoAhPUoaJJSQm\",\n" +
```

sappenin commented on xrpl4j-core/src/test/java/org/xrpl/xrpl4j/model/transactions/AmmClawbackTest.java:
```diff
@@ -0,0 +1,136 @@
+        )
+        .build();
+
+    String json = "{\n" +
```
Separate this into two lines to improve formatter alignment of the rest of the String. E.g., 

```suggestion
    String json = 
      "{\n" +
      ...
```

agutierrez0 replied:
fixed

sappenin commented on xrpl4j-core/src/test/java/org/xrpl/xrpl4j/model/transactions/AmmClawbackTest.java:
```diff
@@ -0,0 +1,136 @@
+        .account(Address.of("rJVUeRqDFNs2xqA7ncVE6ZoAhPUoaJJSQm"))
+        .amount(
+            IssuedCurrencyAmount.builder()
+                .currency("TST")
```
Need to fix all formatting in this class. 2 spaces (no tabs).

agutierrez0 replied:
fixed on this commit: https://github.com/XRPLF/xrpl4j/pull/601/commits/3c03a3a41c9580603de24fe6e6a1fcf6874ea75e

sappenin commented on xrpl4j-core/src/test/java/org/xrpl/xrpl4j/model/transactions/AmmClawbackTest.java:
```diff
@@ -0,0 +1,136 @@
+        "    \"Fee\" : \"10\",\n" +
+        "    \"Sequence\" : 6,\n" +
+        "    \"SigningPubKey\" : \"02356E89059A75438887F9FEE2056A2890DB82A68353BE9C0C0C8F89C0018B37FC\",\n" +
+        "    \"TransactionType\" : \"AMMClawback\"\n" +
```
This test currently fails. Need to add this:

```suggestion
      "  \"TransactionType\" : \"AMMClawback\",\n" +
      "  \"Flags\" : 0" +
```

agutierrez0 replied:
thanks for raising the issues in this unit test file, added the extra field to the json and switched to USD from TST

sappenin commented on xrpl4j-core/src/test/java/org/xrpl/xrpl4j/model/transactions/AmmClawbackTest.java:
```diff
@@ -0,0 +1,136 @@
+                .value("25")
+                .build()
+        )
+        .holder(Address.of("rJVUeRqDFNs2xqA7ncVE6ZoAhPUoaJJSQm"))
```
This test is failing -- This transaction won't build without specifying `asset1` and `asset2`.

agutierrez0 replied:
i have addressed all of the failing tests in this unit test file

sappenin commented on xrpl4j-core/src/test/java/org/xrpl/xrpl4j/model/transactions/AmmClawbackTest.java:
```diff
@@ -0,0 +1,136 @@
+                .value("25")
+                .build()
+        )
+        .holder(Address.of("rJVUeRqDFNs2xqA7ncVE6ZoAhPUoaJJSQm"))
```
This test is failing -- This transaction won't build without specifying `asset1` and `asset2`.

agutierrez0 replied:
fixed all the tests in this unit test file, all 3 passing now

sappenin commented on xrpl4j-core/src/main/resources/definitions.json:
```diff
@@ -1,2775 +1,3026 @@
-        "isSerialized": false,
-        "isSigningField": false,
         "type": "Amount"
       }
```
What version of rippled did you generate this definitions.json from?

agutierrez0 replied:
i initially generated the definitions.json using [this branch](https://github.com/XRPLF/xrpl.js/pull/2826) on the `xrpl.js` library. i compared it to the [definitions.json](https://github.com/XRPLF/xrpl.js/blob/98ef3cb95f70a8c8287a35d84cb4f270656b6c75/packages/ripple-binary-codec/src/enums/definitions.json) that was commited to that repo after adding support for AMMClawback, and noticed mine was different so i used the same one the xrpl.js library used.

should i instead commit the file i get by running the script again on the latest in rippled's develop branch?

agutierrez0 commented on xrpl4j-core/src/test/java/org/xrpl/xrpl4j/model/transactions/AmmClawbackTest.java:
```diff
@@ -0,0 +1,136 @@
+        .signingPublicKey(
+            PublicKey.fromBase16EncodedPublicKey("02356E89059A75438887F9FEE2056A2890DB82A68353BE9C0C0C8F89C0018B37FC")
+        )
+        .flags((AmmClawbackFlags) AmmClawbackFlags.UNSET)
```
i have removed this unnecessary casting, my IDE was also showing i didnt need this

agutierrez0 commented on xrpl4j-core/src/test/java/org/xrpl/xrpl4j/model/transactions/AmmClawbackTest.java:
```diff
@@ -0,0 +1,136 @@
+import org.xrpl.xrpl4j.model.flags.TransactionFlags;
+import org.xrpl.xrpl4j.model.ledger.Issue;
+
+public class AmmClawbackTest extends AbstractJsonTest {
```
i have added this, thanks for mentioning i didn't know these were called a "Javadoc"

nkramer44 commented on xrpl4j-core/src/test/java/org/xrpl/xrpl4j/crypto/signing/SignatureUtilsTest.java:
```diff
@@ -843,6 +844,33 @@ void addSignatureToAmmCreate() {
   }
 
+  @Test
+  void addSignatureToAmmClawback() {
```
Need a test for `signatureUtils.addMultiSignaturesToTransaction`

agutierrez0 replied:
i have added a test for `addMultiSignaturesToTransaction` 👍🏽

nkramer44 commented on xrpl4j-core/src/main/java/org/xrpl/xrpl4j/model/flags/AmmClawbackFlags.java:
```diff
@@ -0,0 +1,51 @@
+/**
+ * {@link TransactionFlags} for {@link AmmClawback} transactions.
+ */
+public class AmmClawbackFlags extends TransactionFlags {
```
Also this class needs a test. Check out `OfferCreateFlagsTests` as an example

agutierrez0 replied:
i have added this static `empty()` method and also added `AmmClawbackFlagTest`. thanks for the references 👍🏽 let me know what you think

nkramer44 commented on xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/AmmIT.java:
```diff
@@ -60,10 +65,11 @@ public class AmmIT extends AbstractIT {
   static boolean shouldNotRun() {
     return System.getProperty("useTestnet") != null ||
-      System.getProperty("useClioTestnet") != null;
+        System.getProperty("useClioTestnet") != null;
```
I think your formatting settings are off, I'll send you the intellij formatting template

agutierrez0 replied:
i have addressed the checkstyle issues i had, latest builds successfully ✅ 

```
[INFO] Reactor Summary for xrpl4j parent HEAD-SNAPSHOT:
[INFO] 
[INFO] xrpl4j parent ...................................... SUCCESS [  1.675 s]
[INFO] xrpl4j :: core ..................................... SUCCESS [ 42.921 s]
[INFO] xrpl4j :: client ................................... SUCCESS [  4.357 s]
[INFO] xrpl4j :: integration-tests ........................ SUCCESS [01:47 min]
[INFO] xrpl4j :: bom ...................................... SUCCESS [  0.044 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:36 min
[INFO] Finished at: 2025-03-10T12:13:12-04:00
[INFO] ------------------------------------------------------------------------
agutierrez in ~/Desktop/xrpl4j on ag/576-amm-clawback λ
```

agutierrez0 commented:
@nkramer44  / @sappenin  - was looking to wrap up this work, could you PTAL? saw comment on [this thread](https://github.com/XRPLF/xrpl4j/pull/556#discussion_r2012392147), not sure if we want to wait till that's merged?

sappenin commented:
> @nkramer44 / @sappenin - was looking to wrap up this work, could you PTAL? saw comment on [this thread](https://github.com/XRPLF/xrpl4j/pull/556#discussion_r2012392147), not sure if we want to wait till that's merged?

@agutierrez0 Please see my code review comments. If they've been addressed, please add a response to let me know to go check. Once addressed, I'll resolve the comments.

agutierrez0 commented on xrpl4j-core/src/test/java/org/xrpl/xrpl4j/model/transactions/AmmClawbackTest.java:
```diff
@@ -0,0 +1,161 @@
+  @Test
+  void testJson() throws JSONException, JsonProcessingException {
+    AmmClawback ammClawback = AmmClawback.builder()
+        .account(Address.of("rJVUeRqDFNs2xqA7ncVE6ZoAhPUoaJJSQm"))
```
I have just applied this formatting to all of the files I have changed, these are my IntelliJ settings:

<img width="990" alt="image" src="https://github.com/user-attachments/assets/9239ebcd-1ffb-438b-a2d3-00485e6c2674" />

let me know if i should change any of these settings and re apply again

agutierrez0 replied:
i have updated this continuation indent and it seems my formatting is now good, i ahve cleaned up some of the files i saw the formatting was wrong. pls let me know if i am missing any other files and thanks for the help here

agutierrez0 commented:
@sappenin  i have addressed your requested changes (format change requested today + changes requested on Feb 25th), could you please take a look?

sappenin reviewed (changes requested):
@agutierrez0 This PR is looking good, and feels _very_ close. I think once we get the formatting worked out, and updates from `main` merged in, there's not much left. 

Also, thanks for your patience in waiting for reviews on this one!

agutierrez0 commented on xrpl4j-core/src/main/java/org/xrpl/xrpl4j/model/transactions/TransactionType.java:
```diff
@@ -338,6 +338,7 @@ public enum TransactionType {
   @Beta
   ORACLE_DELETE("OracleDelete"),
 
+  AMM_CLAWBACK("AMMClawback"),
```
i have updated to include the java doc 👍🏽

agutierrez0 commented on xrpl4j-core/src/main/java/org/xrpl/xrpl4j/crypto/signing/SignatureUtils.java:
```diff
@@ -86,8 +87,8 @@
   private static final SignatureUtils INSTANCE = new SignatureUtils(
-    ObjectMapperFactory.create(),
-    XrplBinaryCodec.getInstance()
+      ObjectMapperFactory.create(),
```
the formatting issues have been fixed after changing the continuation indent to 2 👍🏽

agutierrez0 commented on xrpl4j-core/src/main/java/org/xrpl/xrpl4j/crypto/signing/SignatureUtils.java:
```diff
@@ -117,7 +118,6 @@ public SignatureUtils(final ObjectMapper objectMapper, final XrplBinaryCodec bin
    * Helper method to convert a {@link Transaction} into bytes that can be used directly for signing.
    *
    * @param transaction A {@link Transaction} to be signed.
-   *
```
fixed formatting for this file

agutierrez0 commented on xrpl4j-core/src/main/java/org/xrpl/xrpl4j/crypto/signing/SignatureUtils.java:
```diff
@@ -203,203 +200,206 @@ public UnsignedByteArray toMultiSignableBytes(final Transaction transaction, fin
    */
   public <T extends Transaction> SingleSignedTransaction<T> addSignatureToTransaction(
-    final T transaction, final Signature signature
+      final T transaction, final Signature signature
```
fixed formatting for this file

agutierrez0 commented on xrpl4j-core/src/test/java/org/xrpl/xrpl4j/crypto/signing/SignatureUtilsTest.java:
```diff
@@ -132,16 +133,16 @@ public class SignatureUtilsTest {
-        .build()
-    )
-    .build();
+      .lockingChainDoor(Address.of("rMAXACCrp3Y8PpswXcg3bKggHX76V3F8M4"))
```
fixed formatting for this file as well

agutierrez0 commented on xrpl4j-core/src/main/resources/definitions.json:
```diff
@@ -1,2775 +1,3026 @@
 {
-  "TYPES": {
```
i have synced my fork with the latest and pulled from main 👌🏽

agutierrez0 commented on xrpl4j-core/src/test/resources/data-driven-tests.json:
```diff
@@ -2990,21 +2990,6 @@
       "type": "UInt16",
       "expected_hex": "0005"
     },
```
after pulling the latest from main this indeed did go away

agutierrez0 commented on xrpl4j-integration-tests/src/test/java/org/xrpl/xrpl4j/tests/environment/RippledContainer.java:
```diff
@@ -84,16 +82,16 @@ public class RippledContainer {
    */
   public RippledContainer() {
-    try (GenericContainer<?> container = new GenericContainer<>("rippleci/rippled:2.2.0")) {
+    try (GenericContainer<?> container = new GenericContainer<>("rippleci/rippled:2.3.0")) {
```
yep after pulling the latest from main the changes were gone

agutierrez0 commented:
@sappenin 
> This is a good callout -- there does appear to be a lot of duplicated code. I'd be up for experimenting with how to make this test more concise with less duplicated code. Couple of thoughts:

Yeah I think being able to abstract some of this code for AMMs into function could reduce code duplication which could be nice. yeah would be great to chat, i will try to find a time on your calendar

agutierrez0 commented:
@sappenin  I have addressed your requested changed, could you PTAL again? i finally properly configured my formatting settings and cleaned up files that previously had changes. the PR is a lot cleaner now. pls let me know if anything else needs to be addressed

sappenin commented on xrpl4j-core/src/main/java/org/xrpl/xrpl4j/model/transactions/TransactionType.java:
```diff
@@ -338,6 +338,11 @@ public enum TransactionType {
 
+  /**
+   * The {@link TransactionType} for the {@link AmmClawback} transaction.
+   * is subject to change.</p>
```
```suggestion
```

sappenin reviewed (changes requested):
@agutierrez0 All looks good. Just one small typo to fix in the Javadoc, and then I think we can merge this one.

(If you agree with the change, I think you can commit it directly from Github).

sappenin reviewed (approved):
@agutierrez0 I decided to just merge my suggestion. This is all set to merge. Once the build passes, I'll do that. Thanks for your contribution!!
